### PR TITLE
Adds HorizontalPodAutoscaler resource.

### PR DESCRIPTION
**Pull request by echoboomer**
**What this PR does / why we need it**:
Adds an option for a `HorizontalPodAutoscaler` object to the Buildkite Agent Helm chart. This provides the ability to specify a dictionary called `types` underneath the values heading `horizontalPodAutoscaler` which can optionally be enabled. It is disabled by default.

Out of the box, you can specify `cpu`, `memory`, or `custom` inside this dictionary like this:

```yaml
horizontalPodAutoscaler:
  enabled: true
  maxReplicas: 50
  minReplicas: 10
  types:
    cpu:
      type: Utilization # or AverageValue
      target: 80
    memory:
      type: AverageValue # or Utilization
      target: 80
    custom:
      metricName: requests-per-second
      type: AverageValue # or Value
      target: 2k
```

This is documented in the readme in an additional section.

I also added the settings for `podDisruptionBudget` to the readme because they appeared to be missing.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
This is likely less useful for `cpu` or `memory` scaling, but more useful for `custom` metrics scaling for something like Prometheus. This does not factor in the steps required to expose these metrics within the cluster - it only enables the chart to take advantage of them.

This may need additional testing as I am limited in test environments by the types of external or custom metrics I can use.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by echoboomer** [comment]
Also of note, the `horizontal-pod-autoscaler.yaml` manifest will not render if `types: {}` is left at the default value of an empty dict.

**Comment by rimusz** [comment]
thanks for the PR, I cannot test it myself, maybe @toolmantim or @0x0I can validate it.

**Comment by jack1902** [comment]
@toolmantim @0x0I any chance on this getting looked at soon? (This is of significant interest to our Buildkite Infrastructure, as i'd like to move all agents into our Kube Cluster)

**Comment by O1ahmad** [comment]
Hey got it and sure @jack1902 - will try to take a look today or at least over the weekend.

File: stable/agent/templates/horizontal-pod-autoscaler.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+          apiVersion: apps/v1
+          kind: Deployment
+          name: {{ template "fullname" $ }}
+        target:
+          type: {{ $metricOptions.type }}
+          value: {{ $metricOptions.target }}
+        {{- end }}
+    {{- end }}
+    {{- end }}
+{{- end }}
```
Comment (jack1902) [inline stable/agent/templates/horizontal-pod-autoscaler.yaml]: @echoboomer Wouldn't this be better to allow the caller to specify whatever they want here?

For example, i would want to make use of https://aws.amazon.com/blogs/compute/scaling-kubernetes-deployments-with-amazon-cloudwatch-metrics/

## Why?
Buildkite already have a way to export metrics into CloudWatch (Lambda, Written in Go). This enables me to use things i already have available to me as metrics to scale up and down my Pods and Nodes.

The current implementation here restrics me to the types `Resource` and `Object` but i'd want to use `External`. Whilst you could extend the `if` statements, it feels like it holds too much logic when the caller can just pass in whatever they want for the `HorizontalPodAutoscaler`

File: stable/agent/values.yaml [hunk 2]
```diff
@@ -154,6 +154,14 @@ rbac:
           - create
           - update
 
+horizontalPodAutoscaler:
+  enabled: false
+  # A list of types to use for autoscaling.
+  # Can be cpu, memory, or custom.
```
Comment (jack1902) [inline stable/agent/values.yaml]: See my other comment, but having it limited like this really reduces the usability of this entire feature

**Review by jack1902 — CHANGES_REQUESTED** [review]
Happy to help out where possible on this as this is very close to what i am looking for

**Comment by jack1902** [comment]
@0x0I I've left some comments as i am hoping to move multiple queues into a Kube Cluster via some already existing metrics outside the Cluster within AWS. I have a diagram of the idea that i have which i am happy to share here, as i feel it might be a common pattern followed for numerous Buildkite Agent queues.

Buildkite Kube(1)

File: stable/agent/values.yaml (hunk 2 above)
Comment (echoboomer) [reply stable/agent/values.yaml]: I've resolved this in the latest commit. See below.

File: stable/agent/templates/horizontal-pod-autoscaler.yaml (hunk 1 above)
Comment (echoboomer) [reply stable/agent/templates/horizontal-pod-autoscaler.yaml]: I've resolved this in the latest commit. See below.

**Comment by echoboomer** [comment]
@jack1902 @0x0I I've taken a different approach here based on feedback. This is a bit more simple and leaves it up to the user to decide exactly what is passed in for the `metrics` section. It will only render if `metrics` are actually supplied, and will alert the user if max/min replica counts are not provided. This is documented in the readme. It also lets you override the `apiVersion` for the `hpa` since that is almost entirely a requirement given the spread of use cases.

This approach is a bit more open-ended and isn't as restrictive, and enables someone to do whatever they want with the settings.

**Comment by ianks** [comment]
this is great.

**Comment by hi-artem** [comment]
@jack1902 in your diagram you have worker node group, but isn't [Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) was designed to scale pods?

**Comment by jack1902** [comment]
The agents ran on AWS instances. In EKS, if you consume all of your nodes, you need the cluster autoscaler to scale up your cluster to have additional nodes.

The line between the HPA and cluster autoscaler was there to signify that they work together in order to scale up the number of agents.

I never had the chance to implement this, but I had started to look into running buildkite jobs as kubernetes jobs.

This meant agents only lived as long as job

**Pull request closed without merging**